src/natcap/invest/coastal_blue_carbon/coastal_blue_carbon.py

```diff
@@ -188,941 +194,1521 @@
+                    medium-impact disturbance
+                * ``'high-impact-disturb'`` indicating a
+                    high-impact disturbance
+                * ``''`` (blank), which is equivalent to no carbon change.o
+        biophysical_dict (dict): A ``dict`` mapping of integer landcover codes
+            to biophysical values for disturbance and accumulation values for
+            soil and biomass carbon pools.
 
     Returns:
-        File_Registry (dict): map to collections of output files.
+        Two ``dict``s, each of which maps string keys "soil" and "disturbance"
+        to ``scipy.sparse.dok_matrix`` objects of numpy.float32.  The first
+        such dict contains disturbance magnitudes for the pool for each
+        landcover transition, and the second contains accumulation rates for
+        the pool for the landcover transition.
     """
-    template_raster = C_prior_raster
-
-    T_s_rasters = []
-    A_r_rasters = []
-    E_r_rasters = []
-    N_r_rasters = []
-
-    for snapshot_idx in range(len(snapshot_years)):
-        snapshot_year = snapshot_years[snapshot_idx]
-        T_s_rasters.append(_OUTPUT['carbon_stock'] % (snapshot_year))
-        if snapshot_idx < len(snapshot_years)-1:
-            next_snapshot_year = snapshot_years[snapshot_idx + 1]
-            A_r_rasters.append(_OUTPUT['carbon_accumulation'] % (
-                snapshot_year, next_snapshot_year))
-            E_r_rasters.append(_OUTPUT['cabon_emissions'] % (
-                snapshot_year, next_snapshot_year))
-            N_r_rasters.append(_OUTPUT['carbon_net_sequestration'] % (
-                snapshot_year, next_snapshot_year))
-
-    # Total Net Sequestration
-    N_total_raster = 'total_net_carbon_sequestration.tif'
-
-    raster_registry_dict = {
-        'T_s_rasters': T_s_rasters,
-        'A_r_rasters': A_r_rasters,
-        'E_r_rasters': E_r_rasters,
-        'N_r_rasters': N_r_rasters,
-        'N_total_raster': N_total_raster,
+    table = utils.read_csv_to_dataframe(transition_csv_path, index_col=False)
+
+    lulc_class_to_lucode = {}
+    max_lucode = 0
+    for (lucode, values) in biophysical_dict.items():
+        lulc_class_to_lucode[values['lulc-class']] = lucode
+        max_lucode = max(max_lucode, lucode)
+
+    # Load up a sparse matrix with the transitions to save on memory usage.
+    # The number of possible rows/cols is the value of the maximum possible
+    # lucode we're indexing with plus 1 (to account for 1-based counting).
+    n_rows = max_lucode + 1
+    soil_disturbance_matrix = scipy.sparse.dok_matrix(
+        (n_rows, n_rows), dtype=numpy.float32)
+    biomass_disturbance_matrix = scipy.sparse.dok_matrix(
+        (n_rows, n_rows), dtype=numpy.float32)
+    soil_accumulation_matrix = scipy.sparse.dok_matrix(
+        (n_rows, n_rows), dtype=numpy.float32)
+    biomass_accumulation_matrix = scipy.sparse.dok_matrix(
+        (n_rows, n_rows), dtype=numpy.float32)
```

Comment:
I think I see what you're doing, and I can tell you've thought it through since below you mention the "dict-based approach" from before. But it looks like you're using these just like a dict? To me it's confusing as to why you'd use sparse matrices instead of a dictionary that you index with a tuple. i.e. I wouldn't find `biomass_accumulation_dict[(from_lucode, to_lucode)]` so annoying that I'd use `scipy` matrices. Was there something more complicated before, or is there a larger issue I don't see?

Reply:
Ahh I think I see. Is this so you can use the numpy index indexing on line 1605 below?

Reply:
Yes!  The numpy indexing is extremely convenient, and I also wouldn't be surprised if there's some numpy broadcasting magic happening behind the scenes due to the upfront definition of the sparse matrices' data type.